### PR TITLE
Update prometheus to latest

### LINKED ISSUE #762: dep broken because of moved fsnotify package
```
Solving failure: No versions of gopkg.in/fsnotify.v1 met constraints:
        v1.4.2: unable to update checked out version: fatal: reference is not a tree: 629574ca2a5df945712d3079857300b5e4da0236
: command failed: [git checkout 629574ca2a5df945712d3079857300b5e4da0236]: exit status 128
make: *** [update-vendor] Error 1
```

Merged https://github.com/prometheus/prometheus/pull/3995 to fix upstream.

Will now update in cortex.

### PR DESCRIPTION
**Pull request by aaron7**
- Updated `github.com/prometheus/prometheus` to latest
- Updated `github.com/prometheus/tsdb` to latest
- `make update-vendor`
- Fixed API changes
- Created `querier.Config` for `MaxConcurrent` and `Timeout`

Fixes #762 via prometheus/prometheus#3995 - fsnotify was moved and broke dep here.

### CODE REVIEW

**Comment by bboreham** [comment]
Needs some tweak to compile.

**Comment by bboreham** [comment]
I made those changes.  I did not look more closely whether the underlying Prometheus changes let us remove some shim layers.

**Comment by aaron7** [comment]
I will take a look at test failures now.

File: cmd/querier/main.go [hunk 1]
```diff
@@ -91,13 +92,16 @@ func main() {
 	sampleQueryable := querier.NewQueryable(dist, chunkStore, false)
 	metadataQueryable := querier.NewQueryable(dist, chunkStore, true)
 
-	engine := promql.NewEngine(sampleQueryable, nil)
+	maxConcurrent := 20
+	timeout := 2 * time.Minute
```
Comment (bboreham) [inline cmd/querier/main.go]: Both of these should be command-line parameters.

File: cmd/querier/main.go [hunk 2]
```diff
@@ -91,13 +92,16 @@ func main() {
...
+	engine := promql.NewEngine(util.Logger, nil, maxConcurrent, timeout)
 	api := v1.NewAPI(
 		engine,
 		metadataQueryable,
 		querier.DummyTargetRetriever{},
 		querier.DummyAlertmanagerRetriever{},
 		func() config.Config { return config.Config{} },
+		map[string]string{}, // TODO: include configuration flags
```
Comment (bboreham) [inline cmd/querier/main.go]: ?

**Comment by aaron7** [comment]
@bboreham I think the change you made here: https://github.com/weaveworks/cortex/pull/763/files#diff-bf05432384ae7a97e67ae71efaf68ffcR201 requires slightly different args.

- cfg.NumWorkers - defaults to 1 - promql default engine has 20
- cfg.GroupTimeout - defaults to 10s - promql default engine has 2 minutes

File: cmd/querier/main.go (hunk 1 above)
Comment (aaron7) [reply cmd/querier/main.go]: let's put these in querier config

File: cmd/querier/main.go (hunk 2 above)
Comment (aaron7) [reply cmd/querier/main.go]: this is a recent addition to allow flags to be shown via api endpoint - https://github.com/prometheus/prometheus/pull/3864

We will need to include querier flags here.

**Comment by aaron7** [comment]
Created querier.Config for query engine config and used in `lite`, `querier` and `ruler` with default values of 20 max concurrent queries and 2 minute query timeout.

**Comment by bboreham** [comment]
Why would we particularly want to follow the promql defaults?
The ruler already has its own configuration; we should use it.

File: cmd/querier/main.go (hunk 2 above)
Comment (bboreham) [reply cmd/querier/main.go]: OK; I think we can happily pass blank.

**Comment by aaron7** [comment]
> Why would we particularly want to follow the promql defaults?
> The ruler already has its own configuration; we should use it.

The ruler config has:
- `ruler.num-workers` default: 1 (we override to 10)- Number of rule evaluator worker routines in this process
- `ruler.group-timeout` default: 10s (we override to 30s) - Timeout for rule group evaluation, including sending result to ingester

The new parameters we need to pass:
- number of concurrent promql.Engine queries (prom default is 20)
- timeout for individual queries (prom defaut is 2 mins)

`ruler.num-workers` seems different to `number of concurrent engine queries`. Should we share this?

Is the `ruler.group-timeout` default 10s suitable for query timeouts? We override to 30s, is this suitable? It seems like the answer is yes (there are some issues related to reducing the query timeout).

If we do share these values, should the config names be more generic and not ruler specific?

**Comment by bboreham** [comment]
The number of workers in the ruler is also the maximum number of queries the ruler will issue at the same time. We don’t need another bound. 

The group timeout is an upper bound on the time allowed for a single query; this seems close enough to just use it, in the ruler. 

The querier is different. I can’t see an alternative to adding parameters for it. Maybe unify with the ruler, but removing the ruler-specific ones has to be done carefully.

**Comment by bboreham** [comment]
Is it OK if I pick this up and fix it?

**Comment by aaron7** [comment]
Sorry, other stuff took priority after on-call.

I pushed the change to reuse ruler config - after looking at how the ruler evals groups (sequentially) I agree we should be reusing ruler params.

> Is it OK if I pick this up and fix it?

If there are any other changes to make, sure, please feel free to make them and I can then review.

**Comment by bboreham** [comment]
Thanks!  I added one commit, taking out a couple of TODOs

**Comment by bboreham** [comment]
Couple of odd-looking files under `vendor/github.com/json-iterator/go/` in that last commit.

**Comment by aaron7** [comment]
> Couple of odd-looking files under vendor/github.com/json-iterator/go/ in that last commit.

Yep, these were removed in last commit after `dep ensure` - they shouldn't have been added during the rebase.

**Comment by bboreham** [comment]
Reverted in f2a3355e0bd1f6450bb884ee8fde3475c647ea28 - see #787

**Pull request merged**